agent/src/trident.rs

```diff
@@ -766,7 +782,98 @@ impl DomainNameListener {
     }
 }
 
-pub struct Components {
+pub enum Components {
+    Agent(AgentComponents),
+    #[cfg(target_os = "linux")]
+    Watcher(WatcherComponents),
+}
+
+#[cfg(target_os = "linux")]
+pub struct WatcherComponents {
+    pub api_watcher: Arc<ApiWatcher>,
+    pub libvirt_xml_extractor: Arc<LibvirtXmlExtractor>,
+    pub platform_synchronizer: PlatformSynchronizer,
+    pub domain_name_listener: DomainNameListener,
+    pub running: AtomicBool,
+    tap_mode: TapMode,
+    agent_mode: RunningMode,
+}
+
+#[cfg(target_os = "linux")]
+impl WatcherComponents {
+    fn new(
+        config_handler: &ConfigHandler,
+        stats_collector: Arc<stats::Collector>,
+        session: &Arc<Session>,
+        synchronizer: &Arc<Synchronizer>,
+        exception_handler: ExceptionHandler,
+        remote_log_config: RemoteLogConfig,
+        agent_mode: RunningMode,
+    ) -> Result<Self> {
+        let libvirt_xml_extractor = Arc::new(LibvirtXmlExtractor::new());
+        let candidate_config = &config_handler.candidate_config;
+        let platform_synchronizer = PlatformSynchronizer::new(
+            config_handler.platform(),
+            session.clone(),
+            libvirt_xml_extractor.clone(),
+            exception_handler.clone(),
+            candidate_config.dispatcher.extra_netns_regex.clone(),
+        );
+        let api_watcher = Arc::new(ApiWatcher::new(
+            config_handler.platform(),
+            session.clone(),
+            exception_handler.clone(),
+            stats_collector.clone(),
+        ));
+        let domain_name_listener = DomainNameListener::new(
+            stats_collector.clone(),
+            synchronizer.clone(),
+            remote_log_config,
+            config_handler.static_config.controller_domain_name.clone(),
+            config_handler.static_config.controller_ips.clone(),
+            config_handler.port(),
+        );
+
+        info!("env set ONLY_WATCH_K8S_RESOURCE and IN_CONTAINER, agent only watch k8s resource.");
```

Comment:
when the ONLY_WATCH_K8S_RESOURCE and IN_CONTAINER env are set, agent only watch k8s resource.